File: car-showroom-python/showroom.py
from typing import Optional
from car import Car


class Showroom:
    def __init__(self, name: str):
        if not name or not name.strip():
            raise ValueError("Showroom name cannot be blank")
        self._name = name
        self._inventory: list[Car] = []
        self._next_id = 1

    @property
    def name(self):
        return self._name

    def add_car(self, make: str, model: str, year: int, price: float) -> Car:
        car = Car(self._next_id, make, model, year, price)
        self._next_id += 1
        self._inventory.append(car)
        return car

    def remove_car(self, car_id: int) -> bool:
        for car in self._inventory:
            if car.id == car_id:
                self._inventory.remove(car)
                return True
        return False

    def find_by_id(self, car_id: int) -> Optional[Car]:
        for car in self._inventory:
            if car.id == car_id:
                return car
        return None

    def sell_car(self, car_id: int) -> bool:
        car = self.find_by_id(car_id)
        if car is None:
            return False
        car.available = False
        return True

    def get_all_cars(self) -> list[Car]:
        return list(self._inventory)

    def get_available_cars(self) -> list[Car]:
        return [car for car in self._inventory if car.available]

    def search(self, make: Optional[str] = None, model: Optional[str] = None) -> list[Car]:
        result = self._inventory
        if make is not None:
            result = [car for car in result if car.make == make]
        if model is not None:
            result = [car for car in result if car.model == model]
        return result

    def filter_by_price(self, min_price: float, max_price: float) -> list[Car]:
        if min_price > max_price:
            raise ValueError("min_price cannot exceed max_price")
        return [car for car in self._inventory if min_price <= car.price < max_price]

    def total_cars(self) -> int:
        return len(self._inventory)

    def available_count(self) -> int:
        return sum(1 for car in self._inventory if car.available)

File: car-showroom-java/src/main/java/com/showroom/Showroom.java
package com.showroom;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

public class Showroom {

    private final String name;
    private final List<Car> inventory = new ArrayList<>();
    private int nextId = 1;

    public Showroom(String name) {
        if (name == null || name.isBlank()) throw new IllegalArgumentException("Showroom name cannot be blank");
        this.name = name;
    }

    public String getName() { return name; }

    /** Adds a new car and returns the assigned car (with its generated id). */
    public Car addCar(String make, String model, int year, double price) {
        Car car = new Car(nextId++, make, model, year, price);
        inventory.add(car);
        return car;
    }

    /** Removes a car by id. Returns true if found and removed. */
    public boolean removeCar(int id) {
        return inventory.removeIf(c -> c.getId() == id);
    }

    /** Finds a car by id. */
    public Optional<Car> findById(int id) {
        return inventory.stream().filter(c -> c.getId() == id).findFirst();
    }

    /** Marks a car as sold (unavailable). Returns false if not found or already sold. */
    public boolean sellCar(int id) {
        Optional<Car> car = findById(id);
        if (car.isEmpty() || !car.get().isAvailable()) return false;
        car.get().setAvailable(false);
        return true;
    }

    /** Returns all cars in inventory. */
    public List<Car> getAllCars() {
        return Collections.unmodifiableList(inventory);
    }

    /** Returns only cars that are currently available for sale. */
    public List<Car> getAvailableCars() {
        return inventory.stream().filter(Car::isAvailable).toList();
    }

    /** Case-insensitive search by make and/or model (pass null to skip a filter). */
    public List<Car> search(String make, String model) {
        return inventory.stream()
                .filter(c -> make == null || c.getMake().equalsIgnoreCase(make))
                .filter(c -> model == null || c.getModel().equalsIgnoreCase(model))
                .toList();
    }

    /** Returns cars within an inclusive price range. */
    public List<Car> filterByPrice(double minPrice, double maxPrice) {
        if (minPrice > maxPrice) throw new IllegalArgumentException("minPrice must be <= maxPrice");
        return inventory.stream()
                .filter(c -> c.getPrice() >= minPrice && c.getPrice() <= maxPrice)
                .toList();
    }

    /** Total number of cars in inventory. */
    public int totalCars() { return inventory.size(); }

    /** Number of cars available for sale. */
    public int availableCount() { return (int) inventory.stream().filter(Car::isAvailable).count(); }
}
